File: src/ts/magma/api/collect/Joiner.ts
import { Option } from "../../../magma/api/option/Option";
import { Collector } from "../../../magma/api/collect/Collector";
import { None } from "../../../magma/api/option/None";
import { Some } from "../../../magma/api/option/Some";
export class Joiner implements Collector<string, Option<string>> {
	delimiter: string;
	constructor (delimiter: string) {
		this.delimiter = delimiter;
	}
	static empty(): Joiner {
		return new Joiner("")/*unknown*/;
	}
	createInitial(): Option<string> {
		return new None<string>()/*unknown*/;
	}
	fold(maybe: Option<string>, element: string): Option<string> {
		return new Some<string>(maybe.map((inner: string) => inner + this.delimiter + element/*unknown*/).orElse(element))/*unknown*/;
	}
}

File: src/java/magma/api/collect/Joiner.java
package magma.api.collect;

import magma.api.collect.iter.Collector;

import java.util.Optional;

public record Joiner(String delimiter) implements Collector<String, Optional<String>> {
    @Override
    public Optional<String> createInitial() {
        return Optional.empty();
    }

    @Override
    public Optional<String> fold(Optional<String> maybe, String element) {
        return Optional.of(maybe.map(current -> current + this.delimiter + element)
                .orElse(element));
    }
}
